si5351/si5351.go

```diff
@@ -0,0 +1,443 @@
+package si5351
+
+import (
+	"errors"
+	"math"
+
+	"tinygo.org/x/drivers"
+)
+
+// Device wraps an I2C connection to a SI5351 device.
+type Device struct {
+	bus     drivers.I2C
+	Address uint8
+
+	buf            [8]byte
+	initialised    bool
+	crystalFreq    uint32
+	crystalLoad    uint8
+	pllaConfigured bool
+	pllaFreq       uint32
+	pllbConfigured bool
+	pllbFreq       uint32
+	lastRdivValue  [3]uint8
+}
+
+var errNotInitialised = errors.New("Si5351 not initialised")
+var errInvalidParameter = errors.New("Si5351 invalid parameter")
+
+// New creates a new SI5351 connection. The I2C bus must already be configured.
+//
+// This function only creates the Device object, it does not touch the device.
+func New(bus drivers.I2C) Device {
+	return Device{
+		bus:         bus,
+		Address:     AddressDefault,
+		crystalFreq: CRYSTAL_FREQ_25MHZ,
+		crystalLoad: CRYSTAL_LOAD_10PF,
+	}
+}
+
+// Configure sets up the device for communication
+// TODO error handling
+func (d *Device) Configure() {
+
+	data := d.buf[:1]
+
+	// Disable all outputs setting CLKx_DIS high
+	data[0] = 0xFF
+	d.bus.WriteRegister(d.Address, OUTPUT_ENABLE_CONTROL, data)
+
+	// Set the load capacitance for the XTAL
+	data[0] = d.crystalLoad
+	d.bus.WriteRegister(d.Address, CRYSTAL_INTERNAL_LOAD_CAPACITANCE, data)
+
+	data = d.buf[:8]
+
+	// Power down all output drivers
+	for i := range data {
+		data[i] = 0x80
+	}
+	d.bus.WriteRegister(d.Address, CLK0_CONTROL, data)
+
+	// Disable spread spectrum output.
+	d.DisableSpreadSpectrum()
+
+	d.initialised = true
+
+}
+
+// Connected returns whether a device at SI5351 address has been found.
+func (d *Device) Connected() bool {
+	err := d.bus.Tx(uint16(d.Address), []byte{}, []byte{0})
+	return err == nil
+}
+
+func (d *Device) EnableSpreadSpectrum() (err error) {
+	data := d.buf[:1]
+	err = d.bus.ReadRegister(d.Address, SPREAD_SPECTRUM_PARAMETERS, data)
```

Comment:
The more common practice, at least at google, would be an unnamed error return and then
if err := d.bus.ReadRegister(...); err != nil {
return err
}
...
return d.bus.WriteRegister
it avoids any possible problems around shadowing (which has caused trouble in practice)